File: image_gen.py
#!/usr/bin/env python3
"""
Pixel Art Cat Generator - pi image-generation skill (CPU compatible)
Configurable via environment variables, no interactive prompts
"""
import sys
import os
from pathlib import Path
import time

try:
    import torch
except ImportError:
    print("PyTorch not installed!")
    print("Install: pip install torch -i https://download.pytorch.org/whl/cpu")
    sys.exit(1)

# Auto-detect device
DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Device: {DEVICE}")

try:
    from diffusers import StableDiffusionPipeline
except ImportError as e:
    print(f"Missing package: {e}")
    sys.exit(1)

# Configuration from environment (non-blocking for skill framework)
MODEL_NAME = os.environ.get("MODEL", "runwayml/stable-diffusion-v1-5")
HEIGHT = int(os.environ.get("HEIGHT", 512))
WIDTH = int(os.environ.get("WIDTH", 512))
NUM_STEPS = int(os.environ.get("STEPS", 30))
PROMPT = os.environ.get("PROMPT", "small pixel art picture of a cute fluffy cat, 8bit style")

# Check cache first
CACHE_DIR = Path.home() / ".cache" / "huggingface" / "hub"
MODEL_PATH = CACHE_DIR / "models--runwayml--stable-diffusion-v1-5"
if not MODEL_PATH.exists():
    print(f"Cached model not found at: {MODEL_PATH}")

print("=" * 60)
print("Pixel Art Cat Generator")
print(f"Size: {WIDTH}x{HEIGHT}, Steps: {NUM_STEPS}")
print("=" * 60)

# Load model with CPU/GPU support
try:
    pipe = StableDiffusionPipeline.from_pretrained(
        MODEL_NAME,
        torch_dtype=torch.float16 if DEVICE == "cuda" else torch.float32
    )
    if DEVICE == "cpu":
        pipe.enable_model_cpu_offload()
    print("Model loaded")
except Exception as e:
    print(f"Failed to load model: {e}")
    sys.exit(1)

# Generate images
generator = torch.Generator(device).manual_seed(42)
images = pipe(PROMPT, width=WIDTH, height=HEIGHT, num_inference_steps=NUM_STEPS, generator=generator).images
print(f"Generated {len(images)} image(s)")

# Save outputs
OUTPUT_DIR = Path(__file__).parent / "cat_output"
OUTPUT_DIR.mkdir(exist_ok=True)

for idx, img in enumerate(images):
    path = OUTPUT_DIR / f"cat_{idx:03d}.png"
    img.save(path)
    print(f"Saved: {path}")

print("Done")

File: image_gen.js
#!/usr/bin/env node

/**
 * Image Generation Skill for pi framework
 * Supports FLUX or Stable Diffusion models on CPU/GPU
 */

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

module.exports = async function generateImage() {
  console.log("=".repeat(70));
  console.log("🖼️  Image Generation Skill");
  console.log("=".repeat(70));
  
  // Get configuration from environment or defaults
  const model = process.env.MODEL || "stable-diffusion-v1-5";
  const prompt = process.env.PROMPT || "pixel art cat";
  const numSteps = parseInt(process.env.STEPS) || 30;
  const width = parseInt(process.env.WIDTH) || 512;
  
  // Check Python availability and run generation script
  const scriptPath = path.join(__dirname, '..');
  
  try {
    return await new Promise((resolve, reject) => {
      const pythonScript = fs.existsSync(path.join(scriptPath, 'generate_cat.py')) || 
                           fs.existsSync(path.join(scriptPath, 'generate_cat_final.py')) ?
        `${scriptPath}/generate_cat.py ${prompt} -s${numSteps} -w${width}` :
        '';
      
      if (!pythonScript) {
        console.log("\nNo Python generation script found.");
        resolve({ success: false, error: "No generation script available" });
        return;
      }
      
      console.log(`\n📥 Model: ${model}`);
      console.log(`🎨 Prompt: ${prompt}`);
      console.log(`📐 Size: ${width}x${width}`);
      console.log(`⚙️  Steps: ${numSteps}\n`);
      
      const python = spawn('python', [scriptPath]);
      
      let output = '';
      
      python.stdout.on('data', data => {
        output += data.toString();
        console.log(data.toString());
      });
      
      python.stderr.on('data', data => {
        const err = data.toString();
        if (!err.includes('[Errno 2]')) {
          console.error(err);
        }
      });
      
      python.on('close', code => {
        resolve({
          success: code === 0,
          output
        });
      });
      
      python.on('error', err => {
        reject(new Error(`Python process error: ${err.message}`));
      });
    });
  } catch (error) {
    return { success: false, error: error.message };
  }
};

// Self-execution for testing
if (require.main === module) {
  generateImage()
    .then(result => {
      console.log("=".repeat(70));
      console.log(result.output || result.error);
      process.exit(result.success ? 0 : 1);
    })
    .catch(err => {
      console.error(err.message);
      process.exit(1);
    });
}
